### PR TITLE
imlib: Fix automatic vbuffer count.

### PR DESCRIPTION
**Pull request by iabdalkader**
Before this patch, a 1.5MB framebuffer used a single vbuffer for VGA and smaller resolutions, causing the sensor driver to restart with every snapshot. With this patch, a 1.5MB frame buffer uses 2 vbuffers for VGA/RGB and 3 vbuffers for VGA/grayscale **by default**. In the case of 2 vbuffers, it should leave the rest for `fb_alloc` (not tested), so `fb_alloc` still gets some extra space, but only if there's any left. If you need more space for `fb_alloc`, reserve more if possible.

Additionally, `set_buffers` is now more efficient with buffer sizes. For example, if the source is 1bpp (Bayer) and the destination is 1bpp (grayscale or Bayer), the vbuffer size will be `w*h`, instead of assuming that everything is 2bpp, which allows for more vbuffers.

`set_buffers` now replaces `auto_adjust`. Passing `-1` to `set_buffers` will attempt to use 3 vbuffers, each with a size of `frame_size` if they fit; otherwise, the maximum possible buffers will be used. Passing `1` will use the whole framebuffer. In this case, `frame_size` is ignored.

**Note this definitely needs some more testing and I may have missed some corner cases.**

### CODE REVIEW

File: src/omv/imlib/framebuffer.h [hunk 1]
```diff
@@ -34,6 +34,7 @@ typedef struct framebuffer {
     volatile int32_t tail;
     bool check_head;
     int32_t sampled_head;
+    uint32_t frame_size;
```
Comment (kwagyeman) [inline src/omv/imlib/framebuffer.h:37]: Might want to name this something else since sensor.framesize means something completely different.

File: src/omv/imlib/framebuffer.c [hunk 2]
```diff
@@ -327,19 +312,21 @@ void framebuffer_flush_buffers(bool fifo_flush) {
 
 int framebuffer_set_buffers(int32_t n_buffers) {
     uint32_t total_size = framebuffer_raw_buffer_size();
-    uint32_t size = total_size / n_buffers;
+    uint32_t vbuff_size = (n_buffers == 1) ?
+                          total_size : FB_ALIGN_SIZE_ROUND_UP(framebuffer->frame_size + sizeof(vbuffer_t));
+    int32_t max_buffers = (n_buffers == 1) ? n_buffers : (total_size / vbuff_size);
```
Comment (kwagyeman) [inline src/omv/imlib/framebuffer.c:317]:
```suggestion
    uint32_t vbuff_size = (n_buffers == 1) ?
                          FB_ALIGN_SIZE_ROUND_DOWN(total_size -  sizeof(vbuffer_t)) :
                          FB_ALIGN_SIZE_ROUND_UP(framebuffer->frame_size + sizeof(vbuffer_t));
    int32_t max_buffers = total_size / vbuff_size;
```

I think you need to round down the total size as this doesn't include the vbuffer header which would use some space.

Then max_buffers when n_buffers is one should be the same calculation either way.

File: src/omv/imlib/framebuffer.c [hunk 3]
```diff
@@ -327,19 +312,21 @@ void framebuffer_flush_buffers(bool fifo_flush) {
...
+    uint32_t vbuff_count = (n_buffers == -1) ? IM_MIN(max_buffers, 3) : IM_MIN(max_buffers, n_buffers);
 
-    // Error out if frame buffers are smaller than this...
-    if (size < (sizeof(vbuffer_t) + FRAMEBUFFER_ALIGNMENT)) {
+    if (vbuff_count == 0 || vbuff_size < sizeof(vbuffer_t)) {
```
Comment (kwagyeman) [inline src/omv/imlib/framebuffer.c:320]:
```suggestion
    if (vbuff_count == 0 || vbuff_size < (sizeof(vbuffer_t) + FRAMEBUFFER_ALIGNMENT)) {
```

Should leave the alignment padding as part of this.

**Review by kwagyeman — APPROVED** [review]

File: src/omv/imlib/framebuffer.h (hunk 1 above)
Comment (iabdalkader) [reply src/omv/imlib/framebuffer.h:37]: `raw_buffer_size` is also very misleading. It seems to be the vbuffer buffer size + header but it's called raw.

File: src/omv/imlib/framebuffer.c (hunk 2 above)
Comment (iabdalkader) [reply src/omv/imlib/framebuffer.c:317]: > Then max_buffers when n_buffers is one should be the same calculation either way.

How can `total_size` equal `FB_ALIGN_SIZE_ROUND_DOWN(total_size -  sizeof(vbuffer_t))` for this to work:
```C
     uint32_t vbuff_size = (n_buffers == 1) ?
                          FB_ALIGN_SIZE_ROUND_DOWN(total_size -  sizeof(vbuffer_t)) :
    int32_t max_buffers = total_size / vbuff_size;
```

Comment (iabdalkader) [reply src/omv/imlib/framebuffer.c:317]: I could rearrange things to make that work though. However, vbuff_size here already includes the header size.

File: src/omv/imlib/framebuffer.c (hunk 3 above)
Comment (iabdalkader) [reply src/omv/imlib/framebuffer.c:320]: It's already rounded down/up, right ?

**Comment by iabdalkader** [comment]
```C
uint32_t framebuffer_get_buffer_size() {
    uint32_t size;

    if (framebuffer->n_buffers == 1) {
        // With only 1 vbuffer it's fine to allow the frame buffer size to change given fb_alloc().
        size = framebuffer_raw_buffer_size();
    } else {
        // Whatever the raw size was when the number of buffers were set is locked in.
        size = framebuffer->raw_buffer_size;
    }

```

Shouldn't those be the same now ?

File: src/omv/imlib/framebuffer.h (hunk 1 above)
Comment (iabdalkader) [reply src/omv/imlib/framebuffer.h:37]: `frame_size` doesn't need to be stored, it can be passed as an arg, but I want to keep the function accepting a single arg (count).

File: src/omv/imlib/framebuffer.c (hunk 3 above)
Comment (kwagyeman) [reply src/omv/imlib/framebuffer.c:317]: The vbuffer header yes, but, if the frame buffer storage area itself is 1 byte it will take up 32.

File: src/omv/imlib/framebuffer.c (hunk 2 above)
Comment (kwagyeman) [reply src/omv/imlib/framebuffer.c]: vbuff_size would always be slightly less than total_size so the result would be 1. Which was happening before with (n_buffers == 1) ? n_buffers.

File: src/omv/imlib/framebuffer.h (hunk 1 above)
Comment (kwagyeman) [reply src/omv/imlib/framebuffer.h]: Yeah, I was just suggesting giving it a better name.

File: src/omv/imlib/framebuffer.c [hunk 4]
```diff
@@ -326,24 +309,21 @@ void framebuffer_flush_buffers(bool fifo_flush) {
 }
 
 int framebuffer_set_buffers(int32_t n_buffers) {
-    uint32_t total_size = framebuffer_raw_buffer_size();
-    uint32_t size = total_size / n_buffers;
+    uint32_t avail_size = FB_ALIGN_SIZE_ROUND_DOWN(framebuffer_max_buffer_size());
```
Comment (kwagyeman) [inline src/omv/imlib/framebuffer.c:312]:
```suggestion
    uint32_t avail_size = FB_ALIGN_SIZE_ROUND_DOWN(framebuffer_max_buffer_size() - sizeof(vbuffer_t));
```

One frame buffer still has a vbuffer header that needs to be removed.

File: src/omv/imlib/framebuffer.c (hunk 3 above)
Comment (kwagyeman) [reply src/omv/imlib/framebuffer.c:317]: I guess the frame size could be 0. So... I suppose it's fine.

File: src/omv/imlib/framebuffer.c (hunk 2 above)
Comment (iabdalkader) [reply src/omv/imlib/framebuffer.c]: >vbuff_size would always be slightly less than total_size so the result would be 1

Yes that's right. Well now it's equal so will be 1.

File: src/omv/imlib/framebuffer.h (hunk 1 above)
Comment (iabdalkader) [reply src/omv/imlib/framebuffer.h]: I can't think of better names for this, it is a frame size. However, I'm still thinking about just passing it as a second arg to `set_buffers`. The reason for this is that if you ever call this function with anything other than `1` you must remember to set `frame_size` first, otherwise it will use whatever was there.

**Comment by iabdalkader** [comment]
Assuming this is all good, merging.
